spring-rabbit/src/main/java/org/springframework/amqp/rabbit/listener/DirectMessageListenerContainer.java

```diff
@@ -91,79 +98,243 @@ public void setTaskExecutor(TaskExecutor taskExecutor) {
+				.map(Queue::getName)
+				.collect(Collectors.toList()));
+		super.addQueues(queues);
+	}
+
+	private void addQueues(Collection<String> queues) {
+		if (isRunning()) {
+			synchronized (this.consumersMonitor) {
+				Set<String> current = getQueueNamesAsSet();
+				for (String queue : queues) {
+					if (current.contains(queue)) {
+						throw new IllegalStateException("Queue " + queue + " is already configured for this container: "
+								+ this + ", no queues added");
+					}
+				}
+				List<String> added = new ArrayList<>();
+				for (String queue : queues) {
+					try {
+						consumeFromQueue(queue);
+						added.add(queue);
+					}
+					catch (IOException e) {
+						logger.error("Failed to add queue: " + queue + " undoing adds (if any)");
+						removeQueues(added);
+						throw new AmqpIOException("Failed to add " + queues, e);
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	public boolean removeQueueNames(String... queueNames) {
+		removeQueues(Arrays.asList(queueNames));
+		return super.removeQueueNames(queueNames);
+	}
+
+	@Override
+	public boolean removeQueues(Queue... queues) {
+		removeQueues(Arrays.asList(queues)
+				.stream()
+				.map(Queue::getName)
+				.collect(Collectors.toList()));
+		return super.removeQueues(queues);
+	}
+
+	private void removeQueues(Collection<String> queues) {
+		if (isRunning()) {
+			synchronized (this.consumersMonitor) {
+				for (String queue : queues) {
+					List<SimpleConsumer> consumersForQueue = this.consumersByQueue.remove(queue);
+					if (consumersForQueue != null) {
+						for (SimpleConsumer consumer : consumersForQueue) {
+							try {
+								consumer.getChannel().basicCancel(consumer.getConsumerTag());
+								this.consumers.remove(consumer);
+							}
+							catch (IOException e) {
+								logger.error("Failed to cancel consumer: " + consumer, e);
+							}
+						}
+					}
+				}
+			}
+		}
+	}
+
+	@Override
+	protected boolean canRemoveLastQueue() {
+		return true;
+	}
+
+	private void adjustConsumers(int newCount) {
+		int delta = newCount - this.consumersPerQueue;
+		if (delta != 0) {
+			synchronized (this.consumersMonitor) {
+				if (delta > 0) {
+					List<SimpleConsumer> newConsumers = new ArrayList<>();
+					for (int i = 0; i < delta; i++) {
+						for (String queue : getQueueNames()) {
+							try {
+								newConsumers.add(doConsumeFromQueue(queue));
+							}
+							catch (IOException e) {
+								logger.error("Failed to start a new consumer, reverting", e);
+								for (SimpleConsumer consumer : newConsumers) {
+									try {
+										consumer.getChannel().basicCancel(consumer.getConsumerTag());
+										List<SimpleConsumer> list = this.consumersByQueue.get(queue);
+										if (list != null) {
+											list.remove(consumer);
+										}
+									}
+									catch (IOException e1) {
+										logger.error("Failed to cancel consumer during revert", e);
+									}
+								}
+								throw new AmqpIOException("Failed to increase consumers", e);
```

Comment:
I was thinking because we'd have a container in a strange state (different number of consumers on each queue), but I suppose I can rework it so the counts will be corrected on the next call.